openedx/core/lib/xblock_runtime/id_managers.py

```diff
@@ -0,0 +1,92 @@
+from xblock.runtime import IdReader, IdGenerator
+from opaque_keys.edx.asides import AsideUsageKeyV2, AsideDefinitionKeyV2
+
+from xmodule.modulestore import ModuleStoreEnum
+from xmodule.modulestore.django import modulestore
+
+
+# TODO: this code was copied from x_module.py and modified. Remove it from
+# x_module.py ?
+
+
+class OpaqueKeyReader(IdReader):
+    """
+    IdReader for :class:`DefinitionKey` and :class:`UsageKey`s.
+    """
+    def get_definition_id(self, usage_id):
+        """Retrieve the definition that a usage is derived from.
+
+        Args:
+            usage_id: The id of the usage to query
+
+        Returns:
+            The `definition_id` the usage is derived from
+        """
+        # TODO: if this is a usage v2, then extract the definition from the key
+        # else:
+
+        # TODO: there has got to be a better way:
+        split_modulestore = modulestore().default_modulestore
+        course_entry = split_modulestore._lookup_course(usage_id.course_key.for_branch(ModuleStoreEnum.BranchName.published))
+        return split_modulestore.create_runtime(course_entry, lazy=True).id_reader.get_definition_id(usage_id)
```

Comment:
FWIW, it's probably reasonable to implement this in Split, implemented by getting the structure and doing the block key lookup inside.

More importantly: I'm going to assume compatibility with Old Mongo courses isn't really a consideration?

Reply:
We may not even need this new runtime to support either mongo modulestore, actually. For now I just left this code in because it took me a while to figure out how to get the definition_id from a usage_key outside of the split mongo modulesystem runtime. But it may turn out that this code path in the new runtime is never needed, because it will only be used for things stored in blockstore, where I'm hoping there's a much much much simpler way to get the definition key (see the opaque keys PR).